File: Ordination/Java/src/MergeSort/MergeSort.java
package MergeSort;

import java.util.Arrays;

import sorting.AbstractSort;

public class MergeSort<T extends Comparable<T>> extends AbstractSort<T> {

	@Override
	public void sort(T[] array, int leftIndex, int rightIndex) {
		if(validateParams(array,leftIndex,rightIndex)){
			int mid = (leftIndex + rightIndex) / 2;
			sort(array,leftIndex,mid);
			sort(array,mid + 1, rightIndex);
			merge(array,leftIndex,rightIndex);
		}else{
			//Do Nothing
		}
	}
	private void merge(T[] array,int leftIndex, int rightIndex){
		int mid = (leftIndex + rightIndex) / 2;
		T[] leftArray = Arrays.copyOfRange(array, leftIndex, mid +1);
		T[] rightArray = Arrays.copyOfRange(array, mid + 1, rightIndex + 1);
		
		int i = 0;
		int j = 0;
		while(i< leftArray.length && j < rightArray.length){
			if(leftArray[i].compareTo(rightArray[j]) < 0){
				array[leftIndex++] = leftArray[i++];
			}else{
				array[leftIndex++] = rightArray[j++];
			}
		}
		
		while(i < leftArray.length){
			array[leftIndex++] = leftArray[i++];
		}
		
		while(j < rightArray.length){
			array[leftIndex++] = rightArray[j++];
		}
	}
	
	private boolean validateParams(T[] array,int leftIndex,int rightIndex){
		if(array == null || leftIndex < 0 || rightIndex > array.length - 1 || leftIndex >= rightIndex){
			return false;
		}else{
			return true;
		}
	}
}

File: Python/MergeSort.py

#coding: utf-8

from Sort import Sort

__author__ = "Ionesio Junior"

class MergeSort(Sort):
    ''' Merge sort algorithm implementation '''
    
    def sort(self,array,leftIndex = 0 , rightIndex= None):
        '''Generic Sort
            algorithm.sort(array) -> return complete ordered array
            algorithm.sort(array,leftLimit,rightLimit) -> return partially ordered array
        '''
        if(rightIndex == None):
            rightIndex = len(array) - 1
        if self.__validateParams(array,leftIndex,rightIndex):
            mid = (leftIndex + rightIndex) / 2
            self.sort(array,leftIndex,mid)
            self.sort(array,mid +1,rightIndex)
            self.__merge(array,leftIndex,rightIndex)
    

    def __merge(self,array,leftIndex,rightIndex):
        ''' This method will order and merge two parts of same array '''
        mid = (leftIndex + rightIndex) / 2
        leftArray = array[leftIndex:mid +1]
        rightArray = array[mid +1: rightIndex + 1]
        i = 0
        j = 0
        while(i < len(leftArray) and j < len(rightArray)):
            if(leftArray[i] < rightArray[j]):
                array[leftIndex] = leftArray[i]
                i = i  +1
            else:
                array[leftIndex] = rightArray[j]
                j = j + 1
            leftIndex = leftIndex + 1
    
        while(i < len(leftArray)):
            array[leftIndex] = leftArray[i]
            i = i + 1
            leftIndex = leftIndex + 1
        while(j < len(rightArray)):
            array[leftIndex] = rightArray[j]
            j = j + 1
            leftIndex = leftIndex + 1

    
    def __validateParams(self,array,leftIndex,rightIndex):
        ''' This method verify if parameters are valid,if some parameter isn't valid the method returns false '''    
        if(array == None or leftIndex < 0 or rightIndex > len(array) - 1 or leftIndex >= rightIndex):
            return False
        else:
            return True
